### PR TITLE
[CC-27365] add user_role_grant resource

### PR DESCRIPTION
**Pull request by fantapop**
Previously, the provider supported user role management but only in an authoritative way.  Using the user_role_grants resource would overwrite all roles for a user.  This caused churn for terraform users trying to manage user roles across multiple projects or via the console UI.  We now add a resource, user_role_grant which allows management of a single user role grant.  The existing user_role_grants resource is left to maintain the previous functionality for those that need it.

As part of this changes, I've also renamed the role_resource files which contain the user_role_grants resource after the full name of the role since that seems to be the convention.

**Commit checklist**
- [x] Changelog
- [x] Doc gen (`make generate`)
- [x] Integration test(s)
- [x] Acceptance test(s)
- [x] Example(s)

### CODE REVIEW

File: internal/provider/models.go
Comment (fantapop) [reply internal/provider/models.go:278]: I experimented with this and it doesn't look like we need it.  I did eventually run into an issue with the testing framework where we needed an ID but this was avoided by migrating to the new framework.   See here:
https://developer.hashicorp.com/terraform/plugin/framework/acctests#no-id-found-in-attributes

Here is the PR where I migrated the framework.
https://github.com/cockroachdb/terraform-provider-cockroach/pull/185

Also, I mentioned to you early on that we were planning to add a unique ID for this but decided against it because it didn't fit into our current public api definitions at all.

File: docs/resources/user_role_grant.md
Comment (pawalt) [inline docs/resources/user_role_grant.md:51]: Developer/Admin are not supported

File: docs/resources/user_role_grant.md
Comment (pawalt) [inline docs/resources/user_role_grant.md:65]: Required if resource is cluster/folder is probably useful info.

File: internal/provider/user_role_grant_resource.go
Comment (pawalt) [inline internal/provider/user_role_grant_resource.go:110]: Do we need to do any logic here to check if this assignment is necessary? Or should we just always do it because it's idempotent?

File: internal/provider/user_role_grant_resource.go
Comment (pawalt) [inline internal/provider/user_role_grant_resource.go:251]: Maybe a `uuid.Parse` would be helpful here as a sanity check

File: docs/resources/user_role_grant.md
Comment (fantapop) [reply docs/resources/user_role_grant.md:65]: yes, good point.  Let me know what you think of this:

     ID of the resource. Required if the resource_type is 'FOLDER' or 'CLUSTER'. It should be omitted otherwise.

Additionally, I believe we'll have some docs review before landing this.

File: internal/provider/user_role_grant_resource.go
Comment (fantapop) [reply internal/provider/user_role_grant_resource.go:110]: In general I'm not concerned with idempotency as terraform sprinkles read in various places to determine if something is necessary. Drift can obviously happen and whether it happens during or close to a terraform run is just luck.  It would be detected during the next run and updated.   For this particular case however I can't recall if I've tested what happens when a role grant already exists.  It *should* fail in this case but if our API is idempotent, it may think that it has been able to "adopt" this resource without using import.   I'll do some testing around this.  Maybe I can add an integration test.

File: docs/resources/user_role_grant.md
Comment (fantapop) [reply docs/resources/user_role_grant.md:51]: These are coming through the cloud sdk which gets them from the open api spec which still reflects this. I'll get a PR out right away to remove.  Maybe we can still cherry pick them for this weeks release and I can sneak the change in here.

Comment (fantapop) [reply docs/resources/user_role_grant.md:51]: https://github.com/cockroachlabs/managed-service/pull/16711

File: internal/provider/user_role_grant_resource.go
Comment (fantapop) [reply internal/provider/user_role_grant_resource.go]: I've updated that method and tests to do checking on the uuids.

File: docs/resources/user_role_grant.md
Comment (pawalt) [reply docs/resources/user_role_grant.md]: I like that wording

**Review by pawalt — APPROVED** [review]
LGTM pending docs review

File: internal/provider/user_role_grant_resource.go
Comment (fantapop) [reply internal/provider/user_role_grant_resource.go:111]: So I think this is actually a problem.  If terraform tries to create a resource which already exists and doesn't fail, it means two projects that are accidentally using the same resource will both think they own it.

Comment (fantapop) [reply internal/provider/user_role_grant_resource.go:111]: It could also lead to issues in the same project where a project thinks it has 2 managed resources but only one on the backend.

**Comment by marksoper** [comment]
@mikeCRL @codingconcepts @fantapop @pawalt @shannonbradshaw 
Hi Mike, Rob-
I got word from @shannonbradshaw that you may be able to help with docs for this. A good first step would be to have a look at this PR to see if there are concerns and understand the docs implications.
As you probably know docs are critical for Terraform, as so much Terraform usage begins with `examples`

File: internal/provider/user_role_grant_resource.go
Comment (fantapop) [reply internal/provider/user_role_grant_resource.go:135]: I pushed an update to check for the resource before attempting to "Create" it.  I've also updated the tests to exercise a failure.

**Review by pawalt — APPROVED** [review]
Idempotency changes LGTM

**Comment by fantapop** [comment]
I'm waiting on the new SDK tag to be available before merging this down.  We can add documentation revisions to a later PR if any manifest.  Just let me know.

**Pull request merged**
